design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+`PARTITIONS`/`REPLICAS` If the topic *does not exist* and these values are specified, this command 
+will create a topic with a the set number of partitions and replicas before executing the command. 
+
+If the topic does not exist, and `PARTITIONS`/`REPLICAS` are not supplied, the command will fail
+with the following error message: 
+
+> Topic 'foo' does not exist. If you require an empty topic to be created for the stream/table, 
+> please re-run the statement providing the required 'PARTITION' count and, optional, 'REPLICAS' 
+> count in the WITH clause, e.g. 
+> 
+>     CREATE STREAM BAR (f0, f1) WITH (Kafka_topic='foo', PARTITIONS=4, REPLICAS=2);
+
+If the topic exists and has a different number of partitions/replicas, this command will fail with
+the following error message:
+
+> Cannot issue create command. Topic `<topic name`> exists with different number of partitions or
+> replicas: `<topic description>`.
+
+## Design
+
+### API Design
+
+The above APIs make various trade-offs, outlined below:
+
+* The proposal to use `DELETE` encompasses various trade-offs:
+  * We need a different mechanism to produce tombstones than `INSERT INTO` because there is no good
+  way with the standard SQL to differentiate between an entirely `NULL` value (i.e. tombstone) and 
+  `NULL` values for every column.
+  * Introducing a new syntax (e.g. `INSERT TOMBSTONE`) will require users to understand another
+  language construct and make KSQL depart further from SQL standard
+  * This is a step closer to SQL standard, and we can always support more `DELETE` conditions in
+  the future
+* Allowing `CREATE TABLE/STREAM` to specify a kafka topic that does not yet exist makes it easy to
+use the new APIs without any other system. General sentiment was that it is important for users to
+understand the concept of `PARTITIONS` soon into their journey understanding Kafka. There are four
+alternatives: 
+  * Do not require `PARTITIONS`/`REPLICAS` to create the kafka topic. This is a slight change that
+  will make it easier to experiment, but may open the door to accidentally creating kafka topics.
+  With the release of 5.2, this will also not be consistent with CSAS/CTAS, which default to the
+  source replication/partitioning instead of using some default configured values.
+  * Introduce another `CREATE TOPIC` API which will create empty topics. I believe this will cause
+  confusion, as it departs farther from SQL standard and puts `TOPIC` on the same level as the
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+  * Introduce another `REGISTER TABLE/STREAM` to represent the difference. This moves away from
```

Comment:
I don't think we should so quickly throw away the idea of `REGISTER TABLE/STREAM`. It appears the main objects are:

#### Not standard SQL
It's certainly true that `REGISTER TABLE/STREAM` is not standard SQL, but we must also remember that we're performing an operation for which there is no equivalent in an SQL compliant database. What we're doing is taking raw data stored somewhere, associating a schema and format to it, and making it available for use in KSQL. The equivalent in the DB world would be some SQL command to register some data on the DB server's disk and make it available as a new table.

We have already extended the SQL syntax to allow it to cope with constructs that aren't present in standard SQL DBs, e.g. `CREATE STREAM` and windowed joins, etc.  

#### Adds more complexity for the user to understand.
Adding a new `REGISTER` keyword, (ignoring for a moment that the syntax already allows `REGISTER TOPIC`), does increase the number of keywords, so you could argue it increases the complexity. However, the current CS / CT vs CSAS / CTAS syntax already causes uses a lot of confusion.  The former doesn't actually create anything, while the latter does.  

Personally, I would argue that CS / CT should never of used the `CREATE` keyword, but should have always have been `REGISTER`, because they aren't creating the stream or table: the data already exists. If that were the case, then I think it would be a no-brainer to extend KSQL to support creating empty streams and tables by adding supporting for `CREATE TABLE` and `CREATE STREAM`. 

With the proposed introduction of the new _creating_ semantics for CS / CT statements we have an opportunity to correct the syntax. Yes, there will be a very brief learning curve for existing users when they execute a CT statement and are told the topic already exists, did they mean `REGISTER TABLE`? New users will find the separation between _registering_ existing topics / data in KSQL from  _creating_ new topics / data very intuitive.

Reply:
+1 I strongly support this. We can only get so close to SQL before having to make appropriate concessions, c.f. windowing etc as we do already. 
SQL is so powerful because it's expressive and easy to understand. `REGISTER` aids comprehension; overloading `CREATE STREAM` for three different meanings with additional syntax permutations does not.

Reply:
@hjafarpour @MichaelDrogalis @apurvam - can you chime in with your thoughts?

Reply:
> overloading CREATE STREAM for three different meanings with additional syntax permutations does not.

One thing to note is that after this patch, there will really only be 1 behavior that holds true for all `CREATE STREAM` statements: create if doesn't exist or validate if it does.

Reply:
Also @big-andy-coates, if we change the behavior of `CREATE` to _only_ create new topics, then recovery/restarts will fail with "topic already exists". This means that even if we introduce register, we need `CREATE` to do both creation and validation (not creation).

Reply:
I'm still in favor of only using CS/CT and not introducing new statements (`REGISTER STREAM/TABLE`). I think we can have the following behavior without creating significant confusion. If the topic for CS/CT statement already exists we associate the stream/table with that topic. If the topic does not exist we fail and require user to provide topic properties in the `WITH` clause since we will be creating a new topic. This is a very clear approach and I don't see why it would cause any significant confusion!
As for `REGISTER` vs `CREATE`, I think in addition to the traditional RDBMS systems we should also consider the more recent SQL systems such as Hive, Presto, Impala,... In all of these systems you can create tables on existing files using a `CREATE TABLE` statement.

Reply:
I'm wary of scope creep on this one. I think we should revisit the idea of `register` down the line. I understand that this might be an ideal time from an engineering perspective to address it, but I think we need to make the pragmatic play and continue with what we've got.

Reply:
> Also @big-andy-coates, if we change the behavior of CREATE to only create new topics, then recovery/restarts will fail with "topic already exists". This means that even if we introduce register, we need CREATE to do both creation and validation (not creation).

Yeah, we'd have to preprocess older statements to convert them to REGISTER. No biggie.

> One thing to note is that after this patch, there will really only be 1 behavior that holds true for all CREATE STREAM statements: create if doesn't exist or validate if it does.

Create or validate: that's two behaviours ;)

With regards to @hjafarpour's comments about extending CT/CS not being confusing... as I'm sure @rmoff will chime in, people already find CT/CS confusing because its not creating anything.  He sees this repeatedly in the presentations and demos he does.  And that's before we extend it to optionally create things.

With regards to @MichaelDrogalis comments about scope creep - sure, this is slightly more work. But I think it's wrong to consider this scope creep: all we're doing is defining what we want as the new functionality.  Sure, we could make this change to `REGISTER` later, but that's likely to confuse users even more.

My vote is to make a decision on `CREATE` vs `REGISTER`, i.e. which we want long term, and then implement it now rather than accure tech debt.

Reply:
> Yeah, we'd have to preprocess older statements to convert them to REGISTER. No biggie.

I meant for new statements as well, not just existing ones. Every time someone issues a CREATE we need to convert it to a REGISTER. That seems kind-of weird. Beyond that, headless mode would basically never work with CREATE since it would fail if you restart (that's pretty annoying).

> Create or validate: that's two behaviours ;)

Haha what I meant is that there is one "rule" for all create statements. It might have two parts to it, but there is only one behavior. 

> people already find CT/CS confusing because its not creating anything

I agree that the old behavior is confusing, but now it does create something. If we change it to fail if it exists, then we will be creating different behaviors amongst `C*AS` and `CREATE` statements, which is confusing.

Reply:
> If the topic for CS/CT statement already exists we associate the stream/table with that topic. If the topic does not exist we fail and require user to provide topic properties in the WITH clause since we will be creating a new topic. This is a very clear approach and I don't see why it would cause any significant confusion!

The significant confusion comes when we explain that with CSAS/CTAS the topic _will_ be created even if it _doesn't_ exist (*and* if the user _doesn't_ specify `PARTITION` etc). 

Just because there's a `AS SELECT` appended to the CS/CT, doesn't stop it being a CS/CT is the user's eyes, and thus confusing.

Reply:
> Every time someone issues a CREATE we need to convert it to a REGISTER

Why?  We don't do that for C*AS statements, why would we need to with CREATE?

Reply:
I think an important point to consider here is that `Stream/Table` abstraction is in KSQL layer but `topic` is in Kafka layer. If we think this way, it is very easy to prevent any confusion. In both `CS/CT` and `CSAS/CTAS` we do create new stream/table but these are entities in KSQL layer. When we issue any of the above statements, we do add a NEW stream/table to the metastore! However, each of these statements can have different behavior regarding Kafka topics, they can create a new Kafka topic or they can use an existing one. This is the same for CSAS/CTAS. You can use an existing Kafka topic for the sink in CSAS/CTAS using the `WITH` clause or it will create a new one. However, regardless of existence of Kafka topic, we always have a new Stream/Table in the metastore.
One more point to consider is the relationship between Stream/Table with Kafka topics. Currently we have a one-to-one relation where we have only one topic associated with a stream/table but nothing prevents us from expanding this to many-to-many relation. Indeed this was one of the future extensions I had in mind when I was designing KSQL. So I think as long as we think in terms of Stream/Table in KSQL layer we should be able to explain any confusion that arises.

Reply:
That's a fair point @hjafarpour: both do create entities in the metastore.

> You can use an existing Kafka topic for the sink in CSAS/CTAS using the WITH clause or it will create a new one

I thought it failed if the topic already exists...?

> Currently we have a one-to-one relation where we have only one topic associated with a stream/table

nit: that's not the case, you can register multiple streams / tables using the same underlying topic.

Reply:
I'm +1 on what @big-andy-coates summarized above!

Reply:
@rmoff - thoughts?

Reply:
@rmoff is it still confusing to use CREATE STREAM/CREATE TABLE if we switch to actually creating the streams and tables (as is described in this klip?). I think the proposed behavior is pretty intuitive.